core/include/scipp/core/multi_index.h

```diff
@@ -0,0 +1,97 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2020 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Simon Heybrock
+#pragma once
+
+#include "scipp-core_export.h"
+#include "scipp/core/dimensions.h"
+
+namespace scipp::core {
+
+/// Strides in dataDims when iterating iterDims.
+inline auto get_strides(const Dimensions &iterDims,
+                        const Dimensions &dataDims) {
+  std::array<scipp::index, NDIM_MAX> strides = {};
+  scipp::index d = iterDims.ndim() - 1;
+  for (const auto dim : iterDims.labels()) {
+    if (dataDims.contains(dim))
+      strides[d--] = dataDims.offset(dim);
+    else
+      strides[d--] = 0;
+  }
+  return strides;
+}
+
+template <class... DataDims> class SCIPP_CORE_EXPORT MultiIndex {
+public:
+  constexpr static scipp::index N = sizeof...(DataDims);
+  MultiIndex(const Dimensions &iterDims, const DataDims &... dataDims) {
+    scipp::index d = iterDims.ndim() - 1;
+    for (const auto size : iterDims.shape()) {
+      m_shape[d--] = size;
+      m_end_sentinel *= size;
```

Comment:
Can you clarify why you need this `m_end_sentinel`? Is it for future changes?
I don't see it used apart from in one test.

Reply:
Ah yes, this was intended for handling the subspan indices. It is an end-sentinel, in the sense of C++20 ranges.